src/Nethermind/Nethermind.Core/Collections/ArrayPoolList.cs

```diff
@@ -160,20 +170,23 @@ private bool RemoveAtInternal(int index, bool shouldThrow)
         private bool GuardIndex(int index, bool shouldThrow = true, bool allowEqualToCount = false)
         {
             GuardDispose();
-            if (index < 0)
-            {
-                return shouldThrow
-                    ? throw new ArgumentOutOfRangeException($"Index {index} is below 0.")
-                    : false;
-            }
-            else if (index >= _count && (!allowEqualToCount || index > _count))
+            int count = _count;
+            if ((uint)index > (uint)count || (!allowEqualToCount && index == count))
```

Comment:
can we ignore `index < 0` here?

Reply:
Cast to `uint` means any negative `int`s will become higher than `int.MaxValue` so it gets covered in the greater than check
`(uint)index > (uint)count` as `count` can't be that high

Reply:
Is a common pattern in the runtime to only have a single in range check; without testing for negative values [Span.cs#L148-L159](https://github.com/dotnet/runtime/blob/be84df7642f21702b0e7859f0c134e4d6cfc67eb/src/libraries/System.Private.CoreLib/src/System/Span.cs#L148-L159)

<img width="414" alt="image" src="https://user-images.githubusercontent.com/1142958/218435090-c0930188-864c-4de4-9593-2300205fa779.png">